File: app/src/main/java/vocabletrainer/heinecke/aron/vocabletrainer/lib/Storage/VList.java
package vocabletrainer.heinecke.aron.vocabletrainer.lib.Storage;

import android.os.Parcel;
import android.os.Parcelable;

import java.io.Serializable;
import java.sql.Date;

import static vocabletrainer.heinecke.aron.vocabletrainer.lib.Database.MIN_ID_TRESHOLD;

/**
 * DB Vocable List entry
 */
public class VList implements Parcelable {
    private String nameA;
    private String nameB;
    private String name;
    private Date created;
    private int totalVocs;
    private int unfinishedVocs;
    private int id;

    /**
     * Creates a new VList data object
     *
     * @param id    ID
     * @param nameA Name for A Column
     * @param nameB Name for B Column
     */
    public VList(final int id, final String nameA, final String nameB, final String name, final Date created) {
        this.id = id;
        this.nameA = nameA;
        this.nameB = nameB;
        this.name = name;
        this.created = created;
        this.totalVocs = MIN_ID_TRESHOLD - 1;
        this.unfinishedVocs = MIN_ID_TRESHOLD - 1;
    }

    /**
     * Check whether this entry is existing, according to it's ID<br>
     *     <b>Note:</b> this is not a check whether this entity exists in the Database
     * @return true if the ID is valid
     */
    public boolean isExisting(){
        return isIDValid(id);
    }

    /**
     * Checks whether a given ID is valid, according to MIN_ID_TRESHOLD
     * @param id ID to check
     * @return true if ID is valid
     */
    public static boolean isIDValid(final int id){
        return id >= MIN_ID_TRESHOLD;
    }

    /**
     * Create a new VList with none-ID -1 & current Date
     *
     * @param nameA Name for A Column
     * @param nameB Name for B Column
     * @param name List Name
     */
    public VList(final String nameA, final String nameB, final String name) {
        this(MIN_ID_TRESHOLD - 1, nameA, nameB, name, new Date(System.currentTimeMillis()));
    }

    /**
     * Creates a new table data object, ID only<br>
     * All fields except ID will be empty
     *
     * @param id
     */
    public VList(final int id) {
        this(id, null, null, null,null);
    }

    @Override
    public String toString() {
        return getId() + " " + getName() + " " + getNameA() + " " + getNameB() + " " + getTotalVocs() + " " + getUnfinishedVocs();
    }


    /**
     * Tests for equality based on list ID
     *
     * @param list
     * @return true when list IDs are equal
     */
    public boolean equals(VList list) {
        if (this == list) {
            return true;
        }
        return this.id == list.getId();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof VList) {
            return equals((VList) obj);
        }
        return super.equals(obj);
    }

    /**
     * Returns the amount of vocables this table has<br>
     * The value can be -1 when not set!
     *
     * @return
     */
    public int getTotalVocs() {
        return totalVocs;
    }

    public void setTotalVocs(int total) {
        this.totalVocs = total;
    }

    /**
     * Returns the amount of unfinished vocables<br>
     * The value can be -1 when not set!
     *
     * @return
     */
    public int getUnfinishedVocs() {
        return unfinishedVocs;
    }

    public void setUnfinishedVocs(int unfinished) {
        this.unfinishedVocs = unfinished;
    }

    public String getNameA() {

        return nameA;
    }

    public void setNameA(String nameA) {
        this.nameA = nameA;
    }

    public String getNameB() {
        return nameB;
    }

    public void setNameB(String nameB) {
        this.nameB = nameB;
    }

    public int getId() {
        return id;
    }

    /**
     * Set a new ID
     *
     * @param id new ID
     * @throws IllegalAccessException if a valid ID is already set
     */
    public void setId(int id) {
        if (isIDValid(this.id))
            throw new IllegalAccessError("Can't override existing VList ID");
        else
            this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Date getCreated() {
        return created;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(nameA);
        parcel.writeString(nameB);
        parcel.writeString(name);
        parcel.writeInt(totalVocs);
        parcel.writeInt(unfinishedVocs);
        parcel.writeInt(id);
        parcel.writeLong(created != null ? created.getTime() : 0);
    }

    /**
     * Parcel constructor
     * @param in
     */
    protected VList(Parcel in) {
        nameA = in.readString();
        nameB = in.readString();
        name = in.readString();
        totalVocs = in.readInt();
        unfinishedVocs = in.readInt();
        id = in.readInt();
        created = new Date(in.readLong());
    }

    public static final Creator<VList> CREATOR = new Creator<VList>() {
        @Override
        public VList createFromParcel(Parcel in) {
            return new VList(in);
        }

        @Override
        public VList[] newArray(int size) {
            return new VList[size];
        }
    };
}

File: app/src/main/java/vocabletrainer/heinecke/aron/vocabletrainer/lib/Storage/VList.kt
package vocabletrainer.heinecke.aron.vocabletrainer.lib.Storage

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize
import vocabletrainer.heinecke.aron.vocabletrainer.lib.Database
import java.util.*

/**
 * DB Vocable List entry
 */
@Parcelize
data class VList(
    private var _nameA: String,
    private var _nameB: String,
    private var _name: String,
    var created: Long,
    var changed: Long,

    /**
     * Returns the amount of vocables this table has<br></br>
     * The value can be -1 when not set!
     *
     * @return
     */
    var totalVocs: Int = -1,

    /**
     * Returns the amount of unfinished vocables<br></br>
     * The value can be -1 when not set!
     *
     * @return
     */
    var unfinishedVocs: Int = -1,
    private var _id: Long,
    var uuid: UUID?,
    var isSelected: Boolean = false,
    /**
     * Null if not used
     */
    var categories: MutableList<Category>? = null
) : Parcelable {

    /**
     * Check whether this entry is existing, according to it's ID<br></br>
     * **Note:** this is not a check whether this entity exists in the Database
     * @return true if the ID is valid
     */
    val isExisting: Boolean
        get() = isIDValid(id)

    /**
     * Tests for equality based on list ID
     *
     * @param list
     * @return true when list IDs are equal
     */
    fun equalsId(list: VList?): Boolean {
        return if (this === list) {
            true
        } else id == list!!.id
    }

    fun equalsId(obj: Any?): Boolean {
        return if (obj is VList) {
            equalsId(obj as VList?)
        } else super.equals(obj)
    }

    var name: String
        get() = _name
        set(value) {
            _name = value
            changed = System.currentTimeMillis()
        }

    var nameA: String
        get() = _nameA
        set(value) {
            _nameA = value
            changed = System.currentTimeMillis()
        }

    var nameB: String
        get() = _nameB
        set(value) {
            _nameB = value
            changed = System.currentTimeMillis()
        }

    var id: Long
        get() = _id
        /**
         * Set a new ID
         *
         * @param id new ID
         * @throws IllegalAccessError if a valid ID is already set
         */
        set(value) {
            if (isIDValid(_id)) throw IllegalAccessError("Can't override existing VList ID") else _id =  value
        }

    companion object {
        /**
         * Checks whether a given ID is valid, according to MIN_ID_TRESHOLD
         * @param id ID to check
         * @return true if ID is valid
         */
        @JvmStatic
        fun isIDValid(id: Long): Boolean {
            return id >= Database.MIN_ID_TRESHOLD
        }
        @JvmStatic
        fun isIDValid(id: Int): Boolean {
            return id >= Database.MIN_ID_TRESHOLD
        }

        /**
         * Create a blank VList missing a valid ID and just naming
         */
        fun blank(nameA: String, nameB: String,name: String): VList {
            val time = System.currentTimeMillis()
            return VList(
                _name = name, _nameB = nameB, _nameA = nameA, _id = Database.MIN_ID_TRESHOLD - 1,
                changed = time,
                uuid = null, created = time
            )
        }

        /**
         * Creates a new VList data object
         *
         * @param id    ID
         * @param nameA Name for A Column
         * @param nameB Name for B Column
         */
        fun withId(id: Long, name: String, nameA: String, nameB: String): VList {
            val time = System.currentTimeMillis()
            return VList(
                _name = name, _nameB = nameB, _nameA = nameA, _id = id,
                changed = time,
                uuid = null, created = time
            )
        }
    }
}